controllers/gateway_controller_test.go

```diff
@@ -90,3 +101,348 @@ func TestGatewayReconciler_gatewayHasMatchingGatewayClass(t *testing.T) {
+							AllowedRoutes: &gatewayv1beta1.AllowedRoutes{},
+						},
+					},
+				},
+			},
+			objectsToAdd: []controllerruntimeclient.Object{
+				&corev1.Service{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: "test-namespace",
+						Name:      "service-for-gateway-test-gateway",
+						Labels: map[string]string{
+							gatewayServiceLabel: "test-gateway",
+						},
+					},
+					Spec: corev1.ServiceSpec{
+						Type:      corev1.ServiceTypeLoadBalancer,
+						ClusterIP: "1.1.1.1",
+						Ports: []corev1.ServicePort{
+							{
+								Name:     "udp",
+								Protocol: corev1.ProtocolUDP,
+								Port:     9875,
+							},
+						},
+					},
+					Status: corev1.ServiceStatus{
+						LoadBalancer: corev1.LoadBalancerStatus{
+							Ingress: []corev1.LoadBalancerIngress{
+								{
+									IP: "1.2.3.4",
+								},
+							},
+						},
+					},
+				},
+				&corev1.Endpoints{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "service-for-gateway-test-gateway",
+						Namespace: "test-namespace",
+					},
+				},
+			},
+			testBody: func(t *testing.T, reconciler GatewayReconciler, gatewayReq reconcile.Request, gateway *gatewayv1beta1.Gateway) {
+				ctx := context.Background()
+				_, err := reconciler.Reconcile(ctx, gatewayReq)
+				require.NoError(t, err)
+				newGateway := &gatewayv1beta1.Gateway{}
+				err = reconciler.Client.Get(ctx, gatewayReq.NamespacedName, newGateway)
+				require.NoError(t, err)
+				require.Len(t, newGateway.Status.Addresses, 1)
+				require.Len(t, newGateway.Status.Conditions, 1)
+				require.Equal(t, newGateway.Status.Conditions[0].Status, metav1.ConditionFalse)
+				require.Len(t, newGateway.Status.Listeners, 2)
+				for _, l := range newGateway.Status.Listeners {
+					if l.Name == "http" {
+						require.Len(t, l.SupportedKinds, 0)
+						for _, c := range l.Conditions {
+							if c.Type == string(gatewayv1beta1.ListenerConditionResolvedRefs) {
+								require.Equal(t, c.Status, metav1.ConditionFalse)
+							} else {
+								require.Equal(t, c.Status, metav1.ConditionFalse)
+							}
+						}
+					}
+					if l.Name == "udp" {
+						require.Equal(t, l.SupportedKinds, []gatewayv1beta1.RouteGroupKind{
+							{
+								Group: (*gatewayv1beta1.Group)(&gatewayv1beta1.GroupVersion.Group),
+								Kind:  "UDPRoute",
+							},
+						})
+						for _, c := range l.Conditions {
+							require.Equal(t, c.Status, metav1.ConditionTrue)
+						}
+					}
+				}
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+
+		t.Run(tc.name, func(t *testing.T) {
+			objectsToAdd := []controllerruntimeclient.Object{
+				tc.gatewayClass,
+				tc.gateway,
+			}
+			objectsToAdd = append(objectsToAdd, tc.objectsToAdd...)
+
+			fakeClient := fakectrlruntimeclient.
+				NewClientBuilder().
+				WithScheme(scheme.Scheme).
+				WithObjects(objectsToAdd...).
+				Build()
+
+			reconciler := GatewayReconciler{
+				Client: fakeClient,
+			}
+
+			tc.testBody(t, reconciler, tc.gatewayReq, tc.gateway)
```

Comment:
nit: I would rename `testBody` to `run` personally:

```suggestion
			tc.run(t, reconciler, tc.gatewayReq, tc.gateway)
```